include/swift/AST/LayoutConstraint.h

```diff
@@ -0,0 +1,107 @@
+//===-- LayoutConstraint.h - Layout constraints types and APIs --*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines types and APIs for layout constraints.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_LAYOUT_CONSTRAINT_H
+#define SWIFT_LAYOUT_CONSTRAINT_H
+
+#include "llvm/ADT/StringRef.h"
+
+namespace swift {
+
+class Type;
+class TypeRepr;
+
+/// Describes a layout constraint information.
+enum class LayoutConstraintKind {
+  // It is not a known layout constraint.
+  UnknownLayout,
+  Trivial8 = 1,
```

Comment:
@slavapestov Regarding the more general syntax, I guess something like `_Trivial(N)` would be nice. But currently I expressed these layout constraints as protocols. And we don't have generic protocols yet :-( Otherwise one could just use `_Trivial<N>` instead. To make these constraints parametrized right now, one would need to treat them as something very special, not as protocols. I'm not sure it is a good trade-off. But if we think it is worth it, we could change it.